File: jazzy_ws/src/sli/scripts/isaac_bridge_node.py
#!/usr/bin/env python3
"""
isaac_bridge_node.py

One instance per body.
Subscribes to OrbitState on ~/orbit_state (or ~/cw_state for CW bodies).
Calls /set_entity_state to move the corresponding USD prim in Isaac Sim.

Parameters:
  prim_path       string   USD prim path to move e.g. /World/Sat1
  state_topic     string   topic to subscribe to, default: orbit_state
  attractor_path  string   attractor prim path (to get world position offset)
  scale           float    unit scale factor, default 1.0
                           set to 0.001 if your scene is in meters but
                           orbit state is in km

Note on coordinates:
  OrbitState position is relative to the attractor in km.
  Isaac Sim /set_entity_state expects world-frame meters (or scene units).
  The attractor world position is assumed to be at the origin for now —
  a follow-up can query /get_entity_state for the attractor to handle
  non-origin attractors.
"""

import sys

import rclpy
from rclpy.node import Node

from orbit_interfaces.msg import OrbitState
from simulation_interfaces.srv import SetEntityState
from geometry_msgs.msg import PoseStamped


class IsaacBridgeNode(Node):

    SERVICE_TIMEOUT = 30.0

    def __init__(self):
        super().__init__("isaac_bridge_node")

        self.declare_parameter("prim_path",      "/World/Sat")
        self.declare_parameter("state_topic",    "orbit_state")
        self.declare_parameter("attractor_path", "/World/Earth")
        self.declare_parameter("scale",          1.0)

        self._prim_path      = self.get_parameter("prim_path").value
        self._state_topic    = self.get_parameter("state_topic").value
        self._attractor_path = self.get_parameter("attractor_path").value
        self._scale          = self.get_parameter("scale").value

        # Service client
        self._cli = self.create_client(SetEntityState, "/set_entity_state")

        # Subscriber — topic name is relative so namespace applies
        self._sub = self.create_subscription(
            OrbitState,
            self._state_topic,
            self._on_orbit_state,
            10,
        )

        # Pending call guard — don't stack up calls if Isaac Sim is slow
        self._call_in_flight = False

        self.get_logger().info(
            f"isaac_bridge_node: {self._state_topic} → {self._prim_path} "
            f"scale={self._scale}"
        )

        # Wait for service in background (non-blocking warn only)
        self._service_ready = False
        self._check_service_timer = self.create_timer(
            2.0, self._check_service
        )

    def _check_service(self):
        if self._cli.service_is_ready():
            if not self._service_ready:
                self.get_logger().info("/set_entity_state service ready.")
            self._service_ready = True
            self._check_service_timer.cancel()
        else:
            self.get_logger().warn(
                "/set_entity_state not ready yet — is Isaac Sim running?"
            )

    def _on_orbit_state(self, msg: OrbitState):
        if not self._service_ready:
            return
        if self._call_in_flight:
            return   # drop frame, don't queue up

        # Attractor offset: for now assume attractor is at world origin.
        # TODO: query /get_entity_state for attractor_path to handle
        #       non-origin attractors.
        ax, ay, az = 0.0, 0.0, 0.0

        s = self._scale
        wx = (ax + msg.position.x) * s
        wy = (ay + msg.position.y) * s
        wz = (az + msg.position.z) * s

        req = SetEntityState.Request()
        req.entity = self._prim_path

        req.state.header.stamp    = self.get_clock().now().to_msg()
        req.state.header.frame_id = "world"

        req.state.pose.position.x    = wx
        req.state.pose.position.y    = wy
        req.state.pose.position.z    = wz
        req.state.pose.orientation.w = 1.0
        req.state.pose.orientation.x = 0.0
        req.state.pose.orientation.y = 0.0
        req.state.pose.orientation.z = 0.0

        req.state.twist.linear.x  = msg.velocity.x * s
        req.state.twist.linear.y  = msg.velocity.y * s
        req.state.twist.linear.z  = msg.velocity.z * s

        self._call_in_flight = True
        future = self._cli.call_async(req)
        future.add_done_callback(self._on_set_state_done)

    def _on_set_state_done(self, future):
        self._call_in_flight = False
        try:
            result = future.result()
            if result.result.result_code != 0:
                self.get_logger().warn(
                    f"/set_entity_state failed for {self._prim_path}: "
                    f"code={result.result.result_code} "
                    f"msg={result.result.error_message}"
                )
        except Exception as e:
            self.get_logger().error(f"/set_entity_state exception: {e}")


def main(args=None):
    rclpy.init(args=args)
    node = IsaacBridgeNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: jazzy_ws/src/sli_gnc/src/isaac_bridge_node.cpp
#include <string>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "orbit_interfaces/msg/orbit_state.hpp"
#include "simulation_interfaces/srv/set_entity_state.hpp"

static constexpr uint8_t RESULT_OK = 1;

class IsaacBridgeNode : public rclcpp::Node {
public:
    IsaacBridgeNode() : Node("isaac_bridge_node") {
        declare_parameter("prim_path",      "/World/Sat");
        declare_parameter("state_topic",    "orbit_state");
        declare_parameter("attractor_path", "/World/Earth");
        declare_parameter("scale",          1.0);

        prim_path_   = get_parameter("prim_path").as_string();
        state_topic_ = get_parameter("state_topic").as_string();
        scale_       = get_parameter("scale").as_double();

        cli_ = create_client<simulation_interfaces::srv::SetEntityState>(
            "/set_entity_state");

        sub_ = create_subscription<orbit_interfaces::msg::OrbitState>(
            state_topic_, 10,
            [this](orbit_interfaces::msg::OrbitState::SharedPtr msg) {
                on_orbit_state(msg);
            });

        check_timer_ = create_wall_timer(
            std::chrono::seconds(2),
            [this]() {
                if (cli_->service_is_ready()) {
                    if (!service_ready_) {
                        RCLCPP_INFO(get_logger(),
                            "/set_entity_state ready for %s",
                            prim_path_.c_str());
                        service_ready_ = true;
                        check_timer_->cancel();  
                    }
                } else {
                    RCLCPP_WARN(get_logger(),
                        "/set_entity_state not ready -- is Isaac Sim running?");
                }
            });

        RCLCPP_INFO(get_logger(),
            "isaac_bridge_node: %s -> %s  scale=%.1f",
            state_topic_.c_str(), prim_path_.c_str(), scale_);
    }

private:
    std::string prim_path_;
    std::string state_topic_;
    double      scale_;
    bool        service_ready_ = false;

    rclcpp::Client<simulation_interfaces::srv::SetEntityState>::SharedPtr cli_;
    rclcpp::Subscription<orbit_interfaces::msg::OrbitState>::SharedPtr    sub_;
    rclcpp::TimerBase::SharedPtr check_timer_;

    void on_orbit_state(orbit_interfaces::msg::OrbitState::SharedPtr msg) {
        if (!service_ready_) return;

        auto req = std::make_shared<
            simulation_interfaces::srv::SetEntityState::Request>();

        req->entity                      = prim_path_;
        req->state.header.stamp          = now();
        req->state.header.frame_id       = "world";
        req->state.pose.position.x       = msg->position.x * scale_;
        req->state.pose.position.y       = msg->position.y * scale_;
        req->state.pose.position.z       = msg->position.z * scale_;
        req->state.pose.orientation.w    = msg->attitude.w;
        req->state.pose.orientation.x    = msg->attitude.x;
        req->state.pose.orientation.y    = msg->attitude.y;
        req->state.pose.orientation.z    = msg->attitude.z;
        req->state.twist.linear.x        = msg->velocity.x * scale_;
        req->state.twist.linear.y        = msg->velocity.y * scale_;
        req->state.twist.linear.z        = msg->velocity.z * scale_;

        cli_->async_send_request(req,
            [this](rclcpp::Client<simulation_interfaces::srv::SetEntityState>
                   ::SharedFuture future) {
                try {
                    auto resp = future.get();
                    if (resp->result.result != RESULT_OK) {
                        RCLCPP_WARN(get_logger(),
                            "/set_entity_state failed for %s: code=%d msg=%s",
                            prim_path_.c_str(),
                            resp->result.result,
                            resp->result.error_message.c_str());
                    }
                } catch (const std::exception& e) {
                    RCLCPP_ERROR(get_logger(),
                        "/set_entity_state exception for %s: %s",
                        prim_path_.c_str(), e.what());
                }
            });
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<IsaacBridgeNode>());
    rclcpp::shutdown();
    return 0;
}